src/main/java/org/opensearch/knn/training/TrainingJobRunner.java

```
@@ -0,0 +1,174 @@
+ */
+public class TrainingJobRunner {
+
+    public static Logger logger = LogManager.getLogger(TrainingJobRunner.class);
+
+    private static TrainingJobRunner INSTANCE;
+    private static ModelDao modelDao;
+    private static ThreadPool threadPool;
+
+    private final Semaphore semaphore;
+    private final AtomicInteger jobCount;
+
+    /**
+     * Get singleton instance of TrainingJobRunner
+     *
+     * @return singleton instance of TrainingJobRunner
+     */
+    public static synchronized TrainingJobRunner getInstance() {
+        if (INSTANCE == null) {
+            INSTANCE = new TrainingJobRunner();
+        }
+        return INSTANCE;
+    }
+
+    private TrainingJobRunner() {
+        this.jobCount = new AtomicInteger(0);
+        this.semaphore = new Semaphore(1);
+    }
+
+    /**
+     * Initializes static components.
+     *
+     * @param threadPool threadPool to use to get KNN Training Executor
+     * @param modelDao modelDao used to serialize the models
+     */
+    public static void initialize(ThreadPool threadPool, ModelDao modelDao) {
+        TrainingJobRunner.threadPool = threadPool;
+        TrainingJobRunner.modelDao = modelDao;
+    }
+
+    /**
+     * Execute a training job. This function will first grab a permit, and then serialize the initial model, then
+     * execute training, and then serialize the final result.
+     *
+     * @param trainingJob training job to be executed
+     * @param listener listener to handle final model serialization response (or exception)
+     */
+    public void execute(TrainingJob trainingJob, ActionListener<IndexResponse> listener) throws IOException {
+        // If the semaphore cannot be acquired, the node is unable to execute this job. This allows us to limit
+        // the number of training jobs that enter this function. Although the training threadpool size will also prevent
```

If training threadpool can prevent this, can we avoid semaphore?

---

My intent was to prevent the model from being serialized initially if a job is already running; instead, throw a rejected exception error that will get sent back to user. Submission to the thread pull doesnt happen until after initial serialization. I figured we rather fail fast if we can.